File: pkggo/aiprovider/spec/provider_types.go
package spec

import "context"

// ModelName is an enumeration of model names.
type ModelName string

// ModelInfo represents information about a model.
type ModelInfo struct {
	Name        ModelName    `json:"name"`
	DisplayName string       `json:"displayName"`
	Provider    ProviderName `json:"provider"`
	// MaxPromptLength is set as the one in ModelInfo. The value from input is used if it is less than the ModelInfo
	MaxPromptLength int `json:"maxPromptLength"`
	// MaxOutputLength is set only if it is in input and less than the ModelInfo
	MaxOutputLength int `json:"maxOutputLength"`
	// Add this only if you want to override the one set in provider.
	// Temperature resolution is: ModelInfo > inputParams > ProviderInfo
	DefaultTemperature *float64 `json:"defaultTemperature"`
	// Add this only if you want to override the one set in provider.
	// Streaming resolution is: ModelInfo > ProviderInfo
	StreamingSupport *bool `json:"streamingSupport"`
}

// ProviderName is an enumeration of provider names.
type ProviderName string

// ProviderInfo represents information about a provider.
type ProviderInfo struct {
	Name                     ProviderName           `json:"name"`
	ApiKey                   string                 `json:"apiKey"`
	Engine                   string                 `json:"engine"`
	DefaultOrigin            string                 `json:"defaultOrigin"`
	DefaultModel             ModelName              `json:"defaultModel"`
	AdditionalSettings       map[string]interface{} `json:"additionalSettings"`
	Timeout                  int                    `json:"timeout"`
	ApiKeyHeaderKey          string                 `json:"apiKeyHeaderKey"`
	DefaultHeaders           map[string]string      `json:"defaultHeaders"`
	ChatCompletionPathPrefix string                 `json:"chatCompletionPathPrefix"`
	// Temperature resolution is: inputParams > ModelInfo > ProviderInfo
	DefaultTemperature float64  `json:"defaultTemperature"`
	ModelPrefixes      []string `json:"modelPrefixes"`
	// Streaming resolution is: ModelInfo > ProviderInfo
	StreamingSupport bool `json:"streamingSupport"`
}

func (p *ProviderInfo) IsConfigured() bool {
	return p.ApiKey != ""
}

type CompletionProvider interface {
	MakeCompletion(
		ctx context.Context,
		modelInfo ModelInfo,
		prompt string,
		prevMessages []ChatCompletionRequestMessage,
		inputParams map[string]interface{},
	) (*CompletionRequest, error)
	FetchCompletion(
		ctx context.Context,
		input CompletionRequest,
		onStreamData func(data string) error,
	) (*CompletionResponse, error)
	SetProviderAttribute(
		ctx context.Context,
		apiKey *string,
		defaultModel *string,
		defaultTemperature *float64,
		defaultOrigin *string,
	) error
	IsConfigured(ctx context.Context) bool
}

File: frontend/app/models/aiprovidermodel/provider_types.ts
import { ChatCompletionRequestMessage, CompletionRequest, CompletionResponse } from './chat_types';

export type ModelName = string;
export const DefaultModelName: ModelName = 'gpt-4o';
export const DefaultModelTitle = 'OpenAI GPT 4o';

export type ProviderName = string;
export const DefaultProviderName: ProviderName = 'openai';

export interface ModelInfo {
	name: ModelName;
	displayName: string;
	provider: ProviderName;
	maxPromptLength: number;
	maxOutputLength: number;
	defaultTemperature: number;
}

export interface ProviderInfo {
	name: ProviderName;
	apiKey: string;
	engine: string;
	defaultOrigin: string;
	defaultModel: ModelName;
	additionalSettings: Record<string, any>;
	timeout: number;
	apiKeyHeaderKey: string;
	defaultHeaders: Record<string, string>;
	chatCompletionPathPrefix: string;
	defaultTemperature: number;
	streamingSupport: boolean;
	models: Record<string, ModelInfo>;
	modelPrefixes?: string[];
}

export interface IProviderSetAPI {
	getDefaultProvider(): Promise<ProviderName>;
	setDefaultProvider(provider: ProviderName): Promise<void>;
	getConfigurationInfo(): Promise<Record<string, any>>;
	setAttribute(
		provider: ProviderName,
		apiKey?: string,
		defaultModel?: ModelName,
		defaultTemperature?: number,
		defaultOrigin?: string
	): Promise<void>;

	getCompletionRequest(
		provider: ProviderName,
		prompt: string,
		prevMessages?: Array<ChatCompletionRequestMessage>,
		inputParams?: { [key: string]: any }
	): Promise<CompletionRequest>;
	completion(
		provider: ProviderName,
		input: CompletionRequest,
		onStreamData?: (data: string) => Promise<void>
	): Promise<CompletionResponse | undefined>;
}

export const ProviderInfoDescription = {
	apiKey: 'Your API key for the provider.',
	engine: 'The engine to be used for processing. Is present with Azure etc.',
	defaultOrigin:
		'Default origin to use for requests. This can be used to talk to any server that serves a compatible API',
	defaultModel: 'Default model to use for chat requests',
	additionalSettings: 'Any additional settings to pass to the model. Input as a JSON object',
	timeout: 'The timeout duration in milliseconds.',
	apiKeyHeaderKey: 'The header key for the API key.',
	defaultHeaders: 'The default headers to be included in requests.',
	chatCompletionPathPrefix: 'The path prefix for chat completions.',
	defaultTemperature: 'Default temperature setting for chat requests',
	modelPrefixes: 'Optional prefixes for models.',
};
